de.metas.business/src/main/java/de/metas/customs/CustomsInvoiceService.java

```diff
@@ -0,0 +1,258 @@
+package de.metas.customs;
+
+import java.math.BigDecimal;
+import java.time.LocalDate;
+import java.util.List;
+import java.util.Map;
+
+import org.adempiere.exceptions.AdempiereException;
+import org.compiere.model.I_C_Customs_Invoice;
+import org.compiere.model.I_M_InOutLine;
+import org.compiere.model.I_M_Product;
+import org.compiere.util.Env;
+import org.springframework.stereotype.Service;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import de.metas.bpartner.BPartnerLocationId;
+import de.metas.currency.ICurrencyBL;
+import de.metas.document.DocTypeId;
+import de.metas.document.engine.IDocument;
+import de.metas.document.engine.IDocumentBL;
+import de.metas.document.sequence.IDocumentNoBuilder;
+import de.metas.document.sequence.IDocumentNoBuilderFactory;
+import de.metas.inout.IInOutDAO;
+import de.metas.inout.InOutAndLineId;
+import de.metas.inout.InOutId;
+import de.metas.inout.InOutLineId;
+import de.metas.money.CurrencyId;
+import de.metas.money.Money;
+import de.metas.order.OrderLine;
+import de.metas.order.OrderLineId;
+import de.metas.order.OrderLineRepository;
+import de.metas.product.IProductDAO;
+import de.metas.product.ProductId;
+import de.metas.quantity.Quantity;
+import de.metas.uom.IUOMConversionBL;
+import de.metas.uom.UOMConversionContext;
+import de.metas.uom.UomId;
+import de.metas.user.UserId;
+import de.metas.util.Check;
+import de.metas.util.Services;
+import lombok.NonNull;
+
+/*
+ * #%L
+ * de.metas.business
+ * %%
+ * Copyright (C) 2019 metas GmbH
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program. If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
+
+@Service
+public class CustomsInvoiceService
+{
+
+	private final CustomsInvoiceRepository customsInvoiceRepo;
+	private final OrderLineRepository orderLineRepo;
+
+	public CustomsInvoiceService(
+			@NonNull final CustomsInvoiceRepository customsInvoiceRepo,
+			@NonNull final OrderLineRepository orderLineRepo)
+	{
+		this.customsInvoiceRepo = customsInvoiceRepo;
+		this.orderLineRepo = orderLineRepo;
+
+	}
+
+	public CustomsInvoice generateCustomsInvoice(final BPartnerLocationId bpartnerAndLocationId,
```

Comment:
i believe you should create an immutable (=> `@Value`)request class, instead of having these 6 params

Reply:
done